compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -454,17 +454,38 @@ pub fn impossible_predicates<'tcx>(
     result
 }
 
-fn subst_and_check_impossible_predicates<'tcx>(
+fn instantiated_item_has_impossible_predicates<'tcx>(
     tcx: TyCtxt<'tcx>,
     key: (DefId, SubstsRef<'tcx>),
 ) -> bool {
-    debug!("subst_and_check_impossible_predicates(key={:?})", key);
+    debug!("instantiated_item_has_impossible_predicates(key={:?})", key);
 
     let mut predicates = tcx.predicates_of(key.0).instantiate(tcx, key.1).predicates;
     predicates.retain(|predicate| !predicate.needs_subst());
     let result = impossible_predicates(tcx, predicates);
```

Comment:
It's surprising to me that this doesn't just delegate to `item_has_impossible_predicates_for_item`.

The behavior these two are slightly different; could they be merged?

Reply:
I tried to before, but I had some regressions. 

I think I could try again if I'm smarter with some things though... I'll investigate this weekend.

Reply:
Do you think we should land this in the meantime? If so, I think I'd like a FIXME added, but other than that, r=me

Reply:
It's not ridiculously critical, so I'll play around with it this weekend and either r=you with a better description of why they're different, or unify these two.